src/mrpro/operators/NonUniformFastFourierOp.py

```diff
@@ -0,0 +1,146 @@
+"""Non-Uniform Fast Fourier Operator."""
+
+from collections.abc import Sequence
+
+import numpy as np
+import torch
+from torchkbnufft import KbNufft, KbNufftAdjoint
+
+from mrpro.data.KTrajectory import KTrajectory
+from mrpro.operators.LinearOperator import LinearOperator
+
+
+class NonUniformFastFourierOp(LinearOperator, adjoint_as_backward=True):
+    """Non-Uniform Fast Fourier Operator class."""
+
+    def __init__(
+        self,
+        dim: Sequence[int],
+        recon_matrix: Sequence[int],
+        encoding_matrix: Sequence[int],
+        traj: KTrajectory,
+        nufft_oversampling: float = 2.0,
+        nufft_numpoints: int = 6,
+        nufft_kbwidth: float = 2.34,
+    ) -> None:
+        """Initialize Non-Uniform Fast Fourier Operator.
+
+        Parameters
+        ----------
+        dim
+            dimension along which non-uniform FFT is applied
+        recon_matrix
+            dimension of the reconstructed image corresponding to dim
+        encoding_matrix
+            dimension of the encoded k-space corresponding to dim
+        traj
+            the k-space trajectories where the frequencies are sampled
+        nufft_oversampling
+            oversampling used for interpolation in non-uniform FFTs
+        nufft_numpoints
+            number of neighbors for interpolation in non-uniform FFTs
+        nufft_kbwidth
+            size of the Kaiser-Bessel kernel interpolation in non-uniform FFTs
+        """
+        super().__init__()
+
+        def get_traj(traj: KTrajectory, dims: Sequence[int]):
```

Comment:
now that the NufftOp can be created separately, there are some cases not handled:
What if dim is empty?
What if dim contains invalid axes?
What if dim contains positive axes?

Reply:
I would say dim is empty, the operator should do nothing.
We should add that dim talks about kx ky kz and not k0 k1 k2 here.
So maybe not dim bit direction?

Reply:
also, what about fastfourierop with empty dim?

Reply:
> I would say dim is empty, the operator should do nothing.

added

> also, what about fastfourierop with empty dim?

should be fixed in separate PR

Reply:
> What if dim contains invalid axes?
What if dim contains positive axes?

We could transform them to negative axes based on the dimensions of traj and then verify the result is within (-3,-2,-1)?

Reply:
I started to rewrite the dim logic some time ago to fully separate dim und direction. 
If I remember correctly, what we actuallyean for mifft by dim is NOT dim, but direction.
So I would maybe make it literal kx ky kz ?

Reply:
I guess we have a similar problem with your FFTOp where we also assume that kz ky kx and k2 k1 k0 align, i.e. if I set dim = -1 then I expect to carry out the FFT along k0 with the parameters of (k)x. 

We could either say that FFTOp and NUFFTOp work in k2 k1 k0 space and we have to provide matrix sizes and trajectory and so on in this space. Then we could do the magic of figuring all this out in the FourierOp. This would leave the FFTOp to be more intuitive but would mean that for the NUFFTOp users would have to figure out the traj by themselves. 

Other option would be to provide dim_kzkykx and dim_k2k1k0 to both operators which would make FFTOp less intuitive but mean that NUFFTOp could get the original trajectory as input.

Reply:
I see basically the same two options here:
- nuffOp gets data and omega in shape (batch_different_traj, batch_same_traj, samples), all reshaping is done in FourierOp. 

- NufftOp to gets 'x','y','z' as a `direction` argument. We transform it any ways always to x,y,z where we use it.
This would mean a minimal change to the code.

Not really sure which I prefer, I think a bit the second for know, as it is easier to implement and switch to the first one if it required later on without losing work. Am I missing substantial downsides of the second option?

(Midterm, it would be nice to support in FourierOp also RPE with read-out in z direction, but saved in k0.
But as I said, this is IMHO a separate piece of work for the future)

Reply:
I think Nufft and Fft need k2 k1 k0 because we need to know which dimensions of kdata to combine to samples and which to use as batch. currently we assume this is the same as kz ky kx and hence do the rearrange of kdata in forward/adjoint using nufft_dims which is in kz ky kx space although we are actually rearranging k2 k1 k0 dimensions.

I would prefer it if FftOp and NufftOp used a similar interface. This I think would be much easier to understand by users.

Reply:
Ok, so option two would be to provide direction Literal['x', 'y', 'z',-3,-2,-1] and dim Literal[-3,-2,-1,'k0,'k1',k2'] to nufft.
FFTOp would still only get dim:int. Maye at some point we can add direction here as well  to support the mixed case.

Option one is still all reshaping in FourierOp

Reply:
A bit more work on this topic:
Now `nufft_directions:Sequence[Literal['x', 'y', 'z',-3,-2,-1]]` is provided by the user to define along which directions the nufft should be performed. Based on this ky, kx and kz are selected for Nufft. In a second step it is checked which k0, k1, k2 dimensions ky, kx and kz define (`nufft_dims`). These are then automatically selected. 

In the forward and adjoint this means that the image is permuted/reshaped based on `nufft_directions`, the k-space trajectory and k-space data based on `nufft_dims`. 

This means users don't have to worry about k2-k1-k0-space but can define everything in x-y-z/kx-ky-kz-space. 

It also means that 
> support in FourierOp also RPE with read-out in z direction, but saved in k0

also works now.

The interface is also now more similar to FFT where we either allow SpatialDimension or Sequence[int] for the recon and encoding matrix.